File: apps/web/src/domain/senderClient.test.ts
import { describe, expect, it } from "vitest";
import {
  LOCAL_SENDER_CLIENT,
  isWebSender,
  parseSenderClient,
  senderClientWireField,
  senderOriginNotice,
} from "./senderClient";

describe("sender client origin disclosure", () => {
  it("accepts only the exact canonical wire values", () => {
    expect(parseSenderClient("android")).toBe("android");
    expect(parseSenderClient("web")).toBe("web");
  });

  it("fails closed on missing, mistyped, or unknown values", () => {
    expect(parseSenderClient(undefined)).toBeNull();
    expect(parseSenderClient(null)).toBeNull();
    expect(parseSenderClient("")).toBeNull();
    expect(parseSenderClient("ANDROID")).toBeNull();
    expect(parseSenderClient("Web")).toBeNull();
    expect(parseSenderClient("desktop")).toBeNull();
    expect(parseSenderClient(1)).toBeNull();
    expect(parseSenderClient(true)).toBeNull();
    expect(parseSenderClient(["web"])).toBeNull();
    expect(parseSenderClient({ client: "web" })).toBeNull();
  });

  it("tags this build as the web platform under the stable wire field", () => {
    expect(LOCAL_SENDER_CLIENT).toBe("web");
    expect(senderClientWireField()).toBe("sender_client");
  });

  it("classifies web-origin senders for warning rendering", () => {
    expect(isWebSender("web")).toBe(true);
    expect(isWebSender("android")).toBe(false);
    expect(isWebSender(undefined)).toBe(false);
    expect(isWebSender(null)).toBe(false);
  });

  it("explains each origin without leaking either notice into the other", () => {
    const web = senderOriginNotice("web");
    const android = senderOriginNotice("android");
    expect(web).toMatch(/web client/i);
    expect(web).toMatch(/screenshot/i);
    expect(android).toMatch(/Android app/i);
    expect(web).not.toBe(android);
  });
});

File: android/app/src/test/java/com/abyssal/chat/domain/model/SenderClientTest.kt
package com.abyssal.chat.domain.model

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class SenderClientTest {
    @Test
    fun `accepts only exact canonical wire values`() {
        assertEquals(SenderClient.ANDROID, SenderClient.fromWire("android"))
        assertEquals(SenderClient.WEB, SenderClient.fromWire("web"))
    }

    @Test
    fun `fails closed on missing mistyped or unknown values`() {
        assertNull(SenderClient.fromWire(null))
        assertNull(SenderClient.fromWire(""))
        assertNull(SenderClient.fromWire("ANDROID"))
        assertNull(SenderClient.fromWire("Web"))
        assertNull(SenderClient.fromWire("desktop"))
        assertNull(SenderClient.fromWire(" web"))
        assertNull(SenderClient.fromWire("android "))
    }

    @Test
    fun `wire names round trip through the allowlist`() {
        SenderClient.entries.forEach { client ->
            assertEquals(client, SenderClient.fromWire(client.wireName))
        }
    }

    @Test
    fun `origin notices distinguish the weaker web environment`() {
        val web = SenderClient.WEB.originNotice()
        val android = SenderClient.ANDROID.originNotice()
        assertTrue(web.contains("web client", ignoreCase = true))
        assertTrue(web.contains("screenshot", ignoreCase = true))
        assertTrue(android.contains("Android app", ignoreCase = true))
        assertTrue(web != android)
    }
}
